src/stdio_mgr/types.py

```diff
@@ -0,0 +1,175 @@
+
+    # This is needed to establish a workable MRO.
+
+
+class StdioTupleBase(TupleContextManager, MultiItemTuple):
+    """Base type for a type of stdin, stdout and stderr stream-like objects.
+
+    While it is a context manager, no action is taken on entering or exiting
+    the context.
+
+    Used as a context manager, it will close all of the streams on exit.
+    however it does not open the streams on entering the context manager.
+
+    No exception handling is performed while closing the streams, so any
+    exception occurring the close of any stream renders them all in an
+    unpredictable state.
+    """
+
+    def __new__(cls, iterable):
+        """Instantiate new tuple from iterable containing three streams."""
+        items = list(iterable)
+        assert len(items) == 3, "iterable must be three items"  # noqa: S101
+
+        return super(StdioTupleBase, cls).__new__(cls, items)
+
+    @property
+    def stdin(self):
+        """Return stdin stream."""
+        return self[0]
+
+    @property
+    def stdout(self):
+        """Return stdout stream."""
+        return self[1]
+
+    @property
+    def stderr(self):
+        """Return stderr stream."""
+        return self[2]
+
+
+class ClosingStdioTuple(StdioTupleBase):
+    """Tuple context manager of stdin, stdout and stderr stream-like objects.
+
+    Used as a context manager, it will close all of the streams on exit.
+    however it does not open the streams on entering the context manager.
+
+    No exception handling is performed while closing the streams, so any
+    exception occurring the close of any stream renders them all in an
+    unpredictable state.
+    """
+
+    def close(self):
+        """Close all streams."""
+        return list(self._map("close"))
+
+    def safe_close(self):
+        """Close all streams, ignoring any ValueError."""
+        return list(self.suppress_map(ValueError, "close"))
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Exit context, closing all members."""
+        self.close()
+        return super().__exit__(exc_type, exc_value, traceback)
+
+
+class TextIOTuple(StdioTupleBase):
+    """Tuple context manager of stdin, stdout and stderr TextIOBase objects."""
+
+    _ITEM_BASE = TextIOBase
+
+    # pytest and colorama inject objects into sys.std* that are not real TextIOBase
+    # and fail the assertion of this class
+
+    def __new__(cls, iterable):
+        """Instantiate new tuple from iterable containing three TextIOBase streams."""
+        self = super(TextIOTuple, cls).__new__(cls, iterable)
+        if not self._all(lambda item: isinstance(item, cls._ITEM_BASE)):
+            raise ValueError(
+                "iterable must contain only {}".format(cls._ITEM_BASE.__name__)
+            )
+        return self
+
+
+class FakeIOTuple(StdioTupleBase):
+    """Tuple context manager of stdin, stdout and stderr-like objects."""
+
+
+class AnyIOTuple(StdioTupleBase):
```

Comment:
So...some of the `items` in a `FakeIOTuple` *might* actually be subclasses of `_ITEM_BASE`, but the point of `FakeIOTuple` is to not make any typed promises?

Whereas `TextIOTuple` positively declares all of its `items` as being subclasses of `_ITEM_BASE`?

----

Beyond that, I'm not too much a fan of the name `AnyIOTuple`... it doesn't really reflect how it automatically chooses the type of the return object based on the types of the `items`. Something like `AutoIOTuple` seems better.

Reply:
`AutoIOTuple` is good.

> the point of `FakeIOTuple` is to not make any typed promises?

yup.  `FakeIOTuple` would likely hold the algorithms for managing streams which are dubious, as they are not an `io.TextIOBase`.  We could go one step further and create a class for when all members are `io.IOBase`, before falling back to a name/contract which implies the stream objects are potentially broken.

The one problem with this approach is that often a library doing wrapper/capture will stuff an poorly designed/constructed stream object into `sys.stdin` while building decent classes for `sys.stdout/err`, because they are only interested in the latter.

So there may not be too much benefit in defining "triple" varieties, as there are too many of them.

I expect we would create classes for `WindowsUnbufferedConsole` and `UnixUnbufferedConsole`, etc, if there are noticable implementation details and it is detectable (which it looks like you've solved in the other issue), and they are very common.  Maybe even a `PytestFdCapture`, `Pytest4SysCapture`, `Pytest5SysCapture`, etc if they have noticable differences between versions.